### DISCUSSION TITLE
How to make prop classes overwrite default component classes

### DISCUSSION THREAD

**Discussion by silveltman** [unocss/unocss#3261] [Q&A]
A common thing is creating component with default styles, which optionally take prop classes.
```
// page
<MyComponent className="text-5xl"></MyComponent>

// component
<div className="text-4xl {props.className}"></div>
```

This does not always work, because which class is applied is not decided by the order in which the classes are added to the markup, rather it is decided by the order of the selectors in the generated stylesheet.

What would be the recommended approach here? What I can think of, which does not seem ideal, is the following:

- Add a style tag to the component and use @apply in there
- Add shortcuts for each component

**Comment by chu121su12** [comment]
Personal recommendation: use the `uno-layer-x` variant, say: `uno-layer-component="text-4xl font-italic"`. Add shortcut if necessary (ex. `shrt="text-4xl font-italic"`) that expands to earlier example.

Some explanation: by above, the component will have its rule appliend in the internal layering named `component`. The layer would then be ordered before the default `default` layer which in turn gets overriden with the unprefixed rule if defined later in the page.

**Reply by silveltman** [reply]
Thanks for the reply @chu121su12! Though, I don't fully understand your solution, also not after reading the [docs](https://unocss.dev/config/layers).

What do you mean with `uno-layer-component`, I can not find this feature anywhere. Is this a property I should add to my component? Is this something I should add to my uno config? Is this something I can add the the top of my component like `@unocss-layer-component` similar to `@unocss-include` (which btw I cannot find docs on too).

Could a complete different solution be using [attributify](https://unocss.dev/presets/attributify#installation)? I think it could, if we are able to decide the order in which this preset is added.

Would really appreciate the help!

**Reply by chu121su12** [reply]
I forgot about mentioning atributify. But yes, attributify helps.

Basically, [this layer feature](https://unocss.dev/config/layers#layers).

See [example](https://unocss.dev/play/?html=DwQgtGAEA2CGCeBTATpAxgewLYAcMDtF8AXAfQDcBGScgS1klmOOVoCMBXY2gM3kggA%2BAFDAAJrXKQO%2BDGDhJkYTLgJEyVALwAiYogAexMPuiRaxWNFpptg4AHoJ5EaPBQFKdNjyESFAEw09OhwAM6hAmAi4pIhsOE6MnIeSio%2B6gEAXHqGYP4m0rLyCCjK3mp%2B5P6ZPARGbBjQYrYOTi4xUmhhoTo5RgCsJrYAqvg4yIg8tPqIYpDEGADWRKEAdJAAyoiIkBhcOFzo4ZBsiNAYAO6QtagAwhsbu8hiKLT4AOatkiJAA&config=PTAEFMGcBsEsDsAuBaAJrSBDARtcz5wAPFOQ0eAe2QCcpEbYBjRcVZWAWwAdKbFIAKC69%2BoAN6DQoVOABmCcAGFK8BQHMANFNDc6kcIgCCiBrGwBXRLDkBPbdL1RDASSarID3fsMBVKtoAvqByNJScoADkFlRMkJCRgoLEoogy8pgW0GmyCoQqarDqABSS0jRZUABcoADaOtK1kUwWkIjhtJWRmhKg7tB8NZF0qJGggQC6XlM6kAAWfIgtAjVl0lEtbR3zi8uRQ6wkyNDqoIcofJjw6uCgCwBu4DRV58ismNDdOoFeTgYrdQa3mciH8lGKAEovI4fMZTIxLNY7JDocD-m4PKUgdJIEwPuAagBGAB0ACZUdImKh4EM5qZuJAqiAoJxifNgF91uMoToZoEIYIgA&css=PQKgBA6gTglgLgUzAYwK4Gc4HsC2YDCAyoWABYJQIA0YAhgHYAmYcUD6AZllDhWOqgAOg7nAB0YAGLcwCAB60cggDYIAXGBDAAUKDBi0mXGADe2sGC704AWgDuCGAHNScDQFYADJ4Dc5sAACtMLKAJ5gggCMLPK2ABR2pPBIcsoAlH4WAEa0yADWTlBYqEw2yFjK3Bpw5LxxAOTllVDoYpSMYgAs3vUZ2gC%2BmsBAA&options=N4XyA)

**Comment by silveltman** [comment]
I would like to add the following potential solution:
- Using [attributify](https://unocss.dev/presets/attributify#installation): I think this could, if we are able to decide the order in which this preset is added.
